assembly/src/parsers/ast/mod.rs

```diff
@@ -23,19 +27,98 @@ type ProcMap = BTreeMap<String, ProcedureAst>;
 /// Represents a parsed program AST.
 /// This AST can then be furthur processed to generate MAST.
 /// A program has to have a body and no exported procedures.
-#[derive(Debug)]
+#[derive(Debug, Eq, PartialEq)]
 pub struct ProgramAst {
     pub procedures: ProcMap,
     pub body: Vec<Node>,
 }
 
+impl ProgramAst {
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut program_ast_bytes = ByteWriter::new();
+
+        // procedures
+        program_ast_bytes.write_u16(self.procedures.len() as u16);
+        for (key, value) in self.procedures.iter() {
+            program_ast_bytes
+                .write_string(key)
+                .expect("String serialization failure");
+            value.write_into(&mut program_ast_bytes);
+        }
+
+        // body
+        self.body.write_into(&mut program_ast_bytes);
+
+        program_ast_bytes.get_bytes()
+    }
+
+    pub fn from_bytes(bytes: &mut &[u8]) -> Result<Self, SerializationError> {
+        let mut byte_reader = ByteReader::new(bytes.to_vec());
+
+        let mut procedures = ProcMap::new();
+
+        let num_procedures = byte_reader.get_u16().expect("u16 deserialization failure");
```

Comment:
Here and below, instead of panicking, we should propagate the error - e.g., by doing `let num_procedures = byte_reader.get_u16()?;`